src/expr/src/interpret.rs

```diff
@@ -734,6 +740,14 @@ pub struct ColumnSpecs<'a> {
 }
 
 impl<'a> ColumnSpecs<'a> {
+    // Interpreting a variadic function can lead to exponential blowup: there are up to 4 possibly-
+    // interesting values for each argument (error, null, range bounds...) and in the worst case
+    // we may need to test every combination. We mitigate that here in two ways:
+    // - Adding a linear-time optimization for associative functions like AND, OR, COALESCE.
+    // - Limiting the number of arguments we'll pass on to eval. If this limit is crossed, we'll
+    //   return our default / safe overapproximation instead.
+    const MAX_EVAL_ARGS: usize = 8;
```

Comment:
I'm wondering how the constant value was determined. If I understand correctly, with this value there might be 4^7 = 2^14 = 16384 evaluations at most. This seems probably ok, but I don't really have an accurate intuition about the cost of a single evaluation.

A related question: nowadays, the optimizer runs on a separate thread from the  environmentd main thread. Does the MFP pushdown also run on the optimizer thread? If on the environmentd main thread, then we have to be extra careful here.

(See a similar magic constant [here](https://github.com/MaterializeInc/materialize/blob/6625ba124f12734408c57715338cabf1802b833a/src/transform/src/literal_constraints.rs#L563), where I aimed for at most 10ms per query.)

Reply:
`top` indicates that this work happens in `clusterd` , if I recall correctly. All other queries running on the same compute cluster are blocked for the duration.

Reply:
That's right - this happens effectively just before the persist client would otherwise fetch a part, to decide whether it's safe to skip it.